machinery/gateway_api_topology.go

```diff
@@ -170,6 +256,110 @@ func NewGatewayAPITopology(options ...GatewayAPITopologyOptionsFunc) *Topology {
 		}
 	}
 
+	if o.ExpandGRPCRouteRules {
+		grpcRouteRules := lo.FlatMap(o.GRPCRoutes, GRPCRouteRulesFromGRPCRouteRule)
+		opts = append(opts, WithTargetables(grpcRouteRules...))
+		opts = append(opts, WithLinks(LinkGRPCRouteToGRPCRouteRuleFunc())) // GRPCRoute -> GRPCRouteRule
+
+		if o.ExpandServicePorts {
+			servicePorts := lo.FlatMap(o.Services, ServicePortsFromBackendFunc)
+			opts = append(opts, WithTargetables(servicePorts...))
```

Comment:
This creates the ServicePorts again, after having created within the expansion of HTTPRouteRules. We don't realise that because the URL of the object is the same.

We probably need to rewrite this logic to expand ServicePorts independently from the expansion of route rules, and just link accordingly. IOW, move these two lines out of the expansion of the route rules.

Reply:
Oh, yes, that's it, thanks 👍